libvast/builtins/operators/pseudonymize.cpp

```diff
@@ -0,0 +1,182 @@
+    seed_bytes{};
+  std::vector<std::string> fields;
+
+  template <class Inspector>
+  friend auto inspect(Inspector& f, configuration& x) {
+    return f(x.seed, x.fields);
+  }
+
+  static inline const record_type& layout() noexcept {
+    static auto result = record_type{
+      {"seed", string_type{}},
+      {"fields", list_type{string_type{}}},
+    };
+    return result;
+  }
+};
+
+class pseudonymize_operator : public pipeline_operator {
+public:
+  pseudonymize_operator(configuration config) : config_{std::move(config)} {
+    parse_seed_string();
+  }
+
+  /// Applies the transformation to an Arrow Record Batch with a corresponding
+  /// VAST layout.
+  [[nodiscard]] caf::error
+  add(type layout, std::shared_ptr<arrow::RecordBatch> batch) override {
+    std::vector<indexed_transformation> transformations;
+    for (const auto& field : config_.fields) {
+      for (const auto& index : caf::get<record_type>(layout).resolve_key_suffix(
+             field, layout.name())) {
+        auto transformation = [&](struct record_type::field field,
+                                  std::shared_ptr<arrow::Array> array) noexcept
+          -> std::vector<std::pair<struct record_type::field,
+                                   std::shared_ptr<arrow::Array>>> {
+          if (!caf::holds_alternative<address_type>(field.type)) {
+            VAST_ASSERT(false
+                        && "record batch field to be pseudonymized but does "
+                           "not "
+                           "have address type");
+            VAST_WARN("Field {} is to be pseudonymized but does not contain "
+                      "IP "
+                      "address values; skipping pseudonymization",
+                      field);
+            return {{field, array}};
+          }
+          auto builder
+            = address_type::make_arrow_builder(arrow::default_memory_pool());
+          auto address_view_generator
+            = values(address_type{},
+                     caf::get<type_to_arrow_array_t<address_type>>(*array));
+          for (const auto& address : address_view_generator) {
+            auto append_status = arrow::Status{};
+            if (address) {
+              auto pseudonymized_address
+                = vast::address::pseudonymized(*address, config_.seed_bytes);
+              append_status = append_builder(address_type{}, *builder,
+                                             pseudonymized_address);
+            } else {
+              append_status = builder->AppendNull();
+            }
+            VAST_ASSERT(append_status.ok(), append_status.ToString().c_str());
+          }
+          auto new_array = builder->Finish().ValueOrDie();
+          return {
+            {field, new_array},
+          };
+        };
+        transformations.push_back({index, std::move(transformation)});
+      }
+    }
+    std::sort(transformations.begin(), transformations.end());
```

Comment:
You also need to remove duplicate transformations after the sorting, which is relevant if the users adds multiple extractors that resolve to the same offsets.

Reply:
Fixed in 163cfe3e8
This makes me wonder why `transformations` are supposed to be in a `std:vector` after all...

Reply:
I don't think there's a particular reason for this, so if you want to make this change please go ahead. Such refactorings are always welcome as an initiative (but maybe in a different branch).

Reply:
I'll put the refactoring off to a later date. Maybe there is a valid reason but I can't think of any right now.